core/src/main/scala/com/microsoft/azure/synapse/ml/core/utils/JarLoadingUtils.scala

```diff
@@ -62,4 +57,85 @@ object JarLoadingUtils {
     },
     jarName)
 
+  /**
+    * Get all relevant classes from the ClassLoader.
+    *
+    * Note that the spark ClassPath utility only works for ClassLoaders derived from UrlClassLoader (which sbt uses).
+    * In IntelliJ, the ClassLoader is not a UrlClassLoader (standard App/Ext/Boot loaders), so in that case
+    * we use a different slower method.
+    *
+    * @return A list of classes in the main package and its modules
+    */
+  private def getAllClasses: List[Class[_]] = {
+    // ClassPath is more performant, so we try that first
+    val urlBasedClasses = ClassPath.from(getClass.getClassLoader)
+      .getResources.asScala.toList
+      .map(ri => className(ri.getResourceName))
+      .filter(_.startsWith("com.microsoft.azure.synapse"))
+      .flatMap { cn =>
+        try {
+          Some(Class.forName(cn))
+        } catch {
+          case _: Throwable => None: Option[Class[_]]
+        }
+      }
+
+    // If the list is empty, likely we are running in IntelliJ, so use a different slower method to list the classes
+    if (!urlBasedClasses.isEmpty) urlBasedClasses else getClassesFromPackage("com.microsoft.azure.synapse")
+  }
+
+  /**
+    * Scans all classes accessible from the context class loader which belong to the given package and subpackages.
+    *
+    * @param packageName The base package
+    * @return The classes
+    * @throws ClassNotFoundException
+    * @throws IOException
+    */
+  @throws[ClassNotFoundException]
+  @throws[IOException]
+  private def getClassesFromPackage(packageName: String): List[Class[_]] = {
+    val classLoader = Thread.currentThread.getContextClassLoader
+    assert(classLoader != null)
+
+    // ClassLoader does not expose a class list (except private vars), so scan the jar files directly
+    val path = packageName.replace('.', '/')
+    val resources = classLoader.getResources(path).asIterator()
+    val dirs = new ListBuffer[File]
+    for (resource <- resources.asScala) {
+      dirs += new File(resource.getFile)
+    }
+
+    val classes = new ListBuffer[Class[_]]
+    for (directory <- dirs) {
+      classes ++= findClassesInDirectory(directory, packageName)
+    }
+    classes.toList
+  }
+
+  /**
+    * Recursive method used to find all classes in a given directory and subdirs.
+    *
+    * @param directory   The base directory
+    * @param packageName The package name for classes found inside the base directory
+    * @return The classes
+    * @throws ClassNotFoundException
+    */
+  @throws[ClassNotFoundException]
+  private def findClassesInDirectory(directory: File, packageName: String): ListBuffer[Class[_]] = {
+    val classes = new ListBuffer[Class[_]]
```

Comment:
can convert to map chaining or a for yield statement if you want to keep imperative paradigm without mutability

Reply:
https://alvinalexander.com/scala/scala-for-loop-yield-examples-yield-tutorial/